lib/kube/proxy/forwarder.go

```diff
@@ -691,8 +691,14 @@ func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUs
 	}
 
 	disconnectExpiredCert := roles.AdjustDisconnectExpiredCert(authPref.GetDisconnectExpiredCert())
-	if !certExpires.IsZero() && disconnectExpiredCert {
-		authCtx.disconnectExpiredCert = certExpires
+	if disconnectExpiredCert {
+		if !identity.MFAVerifiedSessionExpires.IsZero() {
+			// Cause MFA verified sessions to disconnect on issuing certs expiry
+			// (see https://github.com/gravitational/teleport/issues/18544).
+			authCtx.disconnectExpiredCert = identity.MFAVerifiedSessionExpires
+		} else if !certExpires.IsZero() {
+			authCtx.disconnectExpiredCert = certExpires
+		}
```

Comment:
I need some guidance here. I suspect this logic is wrong however I know next to nothing about how kubernetes access works in general or what this section is doing in particular.

The typical process for setting `disconnectExpiredCert` is encoded in `GetDisconnectExpiredCertFromIdentity`
https://github.com/gravitational/teleport/blob/6263b81de3ea284c6b93746a7c56a0ae954496f2/lib/srv/monitor.go#L429-L457

However because we are using a special value `certExpires` here, rather than the typical case of using `identity.Expires`, I suspect that using `identity.MFAVerifiedSessionExpires` on line 698 is wrong. Probably I instead want to parse the cert that `certExpires` come from and use the new extension (see top level description) that corresponds to `MFAVerifiedSessionExpires`, but I need somebody who actually understands the system to confirm.

Tagging those who `git blame` tells me have worked on code around here: @dboslee @tigrato @Joerger @smallinsky @klizhentas

Reply:
Here's what I think it should actually look like, still please somebody double check that this makes sense in context:

919aca472825265ea285559e2ec2fe544c4564aa

Reply:
Kubernetes Proxy caches users' certificates and uses the `certExpires` (previously not after) to differentiate certificates for the same user (they will end up with different keys in the cache). This is important for cases where the user has connected to a proxy with a certificate that missed some Traits and has to re-login to regain them (check #18070 for more details)

As far as I can see, `Expires` is not marshaled into the certificate and it's not carried around. It depends on what is passed as the second arg to `tlsca.FromSubject` function.

```go
	id, err := tlsca.FromSubject(csr.Subject, time.Time{})
// or
        id, err := tlsca.FromSubject(csr.Subject,csr.NotAfter)
```
 and we have cases where we pass a zero time [[1](https://github.com/gravitational/teleport/blob/e17bcb13014e7b71bf787b3fcab5e8bd754d3b95/lib/auth/db.go#L152) ,[2](https://github.com/gravitational/teleport/blob/e17bcb13014e7b71bf787b3fcab5e8bd754d3b95/lib/auth/kube.go#L99)].

Since you pass the `clientCert.NotAfter` as a second arg to `tlsca.FromSubject`, `Expires` will be populated correctly.

When disconnect on cert expiration is enabled, the `disconnectExpiredCert` should be the cert expiration date if MFA is disabled or the `identity.MFAVerifiedSessionExpires` when MFA is enabled. 

From what I saw, your changes are correct and working accordingly.

Reply:
Thanks for testing